### PR TITLE
grndb recover: close used objects immediately

### PR DESCRIPTION
**Pull request by komainu8**
We can reduce memory usage by this.
This may decrease performance but it will be acceptable.

Note that "grndb check" doesn't close used objects immediately yet.

### CODE REVIEW

**Comment by kou** [comment]
This approach will not work for objects that refer other objects such as an index column. An index column refers its lexicon.

We should use `grn_ctx_push_temporary_open_space()`/`grn_ctx_pop_temporary_open_space()` like `grn_db_recover()` does.

File: lib/mrb/mrb_object.c [hunk 1]
```diff
@@ -254,6 +254,32 @@ grn_mrb_object_is_closed(mrb_state *mrb, mrb_value self)
   return mrb_bool_value(object == NULL);
 }
 
+mrb_value
+grn_mrb_object_push_open_space(mrb_state *mrb, mrb_value self)
```
Comment (kou) [inline lib/mrb/mrb_object.c:258]:
```suggestion
grn_mrb_object_push_temporary_open_space(mrb_state *mrb, mrb_value self)
```

File: lib/mrb/mrb_object.c [hunk 2]
```diff
@@ -254,6 +254,32 @@ grn_mrb_object_is_closed(mrb_state *mrb, mrb_value self)
...
+{
+  grn_ctx *ctx = (grn_ctx *)mrb->ud;
+  grn_bool is_close_opened_object_mode = (grn_thread_get_limit() == 1);
```
Comment (kou) [inline lib/mrb/mrb_object.c:261]: We should not do this in this function.
It should be done by caller.

File: lib/mrb/scripts/command_line/grndb.rb [hunk 3]
```diff
@@ -603,8 +603,13 @@ def recover
           logger.log(:info, "Recovering database: <#{@database.path}>")
           if @force_truncate
             @database.each do |object|
```
Comment (kou) [inline lib/mrb/scripts/command_line/grndb.rb:605]: How about the following API?

```suggestion
            @database.each(use_temporary_open_space: true) do |object|
```

File: lib/mrb/mrb_object.c [hunk 4]
```diff
@@ -254,6 +254,32 @@ grn_mrb_object_is_closed(mrb_state *mrb, mrb_value self)
...
+
+  if (is_close_opened_object_mode) {
+    grn_ctx_push_temporary_open_space(ctx);
```
Comment (kou) [inline lib/mrb/mrb_object.c:264]: We should call `grn_mrb_ctx_check(mrb);` after calling any Groonga API.

File: lib/mrb/scripts/database.rb [hunk 5]
```diff
@@ -38,7 +38,11 @@ def each(options=nil)
       context = Context.instance
       each_raw(options) do |id, cursor|
         object = context[id]
-        yield(object) if object
+        if object
+          object.push_temporary_open_space if options[:use_temporary_open_space]
+          yield(object)
+          object.pop_temporary_open_space if options[:use_temporary_open_space]
+        end
```
Comment (kou) [inline lib/mrb/scripts/database.rb:45]:
```ruby
each_raw(options) do |id, cursor|
  if options[:use_temporary_open_space]
    context.push_temporary_open_space
    begin
      object = context[id]
      yield(object) if object
    ensure
      context.pop_temporary_open_space
    end
  else
    object = context[id]
    yield(object) if object
  end
end

File: lib/mrb/scripts/command_line/grndb.rb [hunk 6]
```diff
@@ -602,7 +602,7 @@ def initialize(output)
         def recover
           logger.log(:info, "Recovering database: <#{@database.path}>")
           if @force_truncate
-            @database.each do |object|
+            @database.each(use_temporary_open_space: close_opened_object_mode?) do |object|
```
Comment (kou) [inline lib/mrb/scripts/command_line/grndb.rb:605]: We can assume that `grndb` doesn't use multiple threads.

File: lib/mrb/mrb_ctx.c [hunk 7]
```diff
@@ -832,6 +832,28 @@ grn_mrb_ctx_to_exception(mrb_state *mrb)
   return mrb_exc_new_str(mrb, error_class, mrb_str_new_cstr(mrb, message));
 }
 
+mrb_value
```
Comment (kou) [inline lib/mrb/mrb_ctx.c:835]: This should not be exported:

```suggestion
static mrb_value
```

Comment (kou) [reply lib/mrb/mrb_ctx.c:835]: They should be defined right after `ctx_is_opened()`.

**Comment by kou** [comment]
Can you add tests for this case?

**Comment by komainu8** [comment]
Tests of `grndb recover ----force-truncate` and `grndb recover --force-lock-clear` have already exist in https://github.com/groonga/groonga/blob/master/test/command_line/suite/grndb/test_recover.rb#L177-L205 and https://github.com/groonga/groonga/blob/master/test/command_line/suite/grndb/test_recover.rb#L358-L380.

Are these tests not enough?

**Comment by kou** [comment]
If we can reuse existing tests for this case, the existing tests should be failed. Because this change should close more objects.
If the existing tests aren't failed, this case isn't covered by existing tests.

**Comment by komainu8** [comment]
Groonga doesn't output a part of the log(`open existing file ...`) on Linux. Therefore, existing tests don't fail on Linux, but they fail on Windows.
Therefore, I modified expected values for Windows of existing tests.

**Comment by kou** [comment]
`DUMP` log level dumps open/close logs on Linux.
BTW, why do you want to check open log not close log?
Does this change close objects immediately?

**Comment by kou** [comment]
`grndb` wasn't ran in single thread mode.

**Comment by kou** [comment]
I'll merge this because Linux test failures should be unrelated.
This changes only grndb.

**Pull request merged**